From c5bd47ceea8d229f7bf35f84e6743ec7d5d50592 Mon Sep 17 00:00:00 2001
From: Martin Owens <doctormo@geek-2.com>
Date: Fri, 1 Jul 2022 16:22:19 -0400
Subject: [PATCH] Add option to have snap toolbar be permanent

This adds an option to have single-click vertical toolbar
for the snap menu. It adds these using the minimum amount of
code and glade file to keep everything consistant.
---
 share/ui/toolbar-snap.ui               | 415 ++++++++++++++++++++++++-
 src/ui/dialog/inkscape-preferences.cpp |   3 +-
 src/widgets/desktop-widget.cpp         |  25 +-
 src/widgets/desktop-widget.h           |   2 +
 src/widgets/toolbox.cpp                |   1 -
 5 files changed, 441 insertions(+), 5 deletions(-)

diff --git a/share/ui/toolbar-snap.ui b/share/ui/toolbar-snap.ui
index 3703c09922..ba7efa7ce6 100644
--- a/share/ui/toolbar-snap.ui
+++ b/share/ui/toolbar-snap.ui
@@ -708,9 +708,9 @@
           <object class="GtkSeparator" id="indent-space">
             <property name="visible">True</property>
             <property name="can-focus">False</property>
+            <property name="opacity">0</property>
             <property name="margin-start">5</property>
             <property name="margin-end">5</property>
-            <property name="opacity">0</property>
           </object>
           <packing>
             <property name="left-attach">0</property>
@@ -1134,6 +1134,12 @@
         <child>
           <placeholder/>
         </child>
+        <child>
+          <placeholder/>
+        </child>
+        <child>
+          <placeholder/>
+        </child>
       </object>
     </child>
   </object>
@@ -1157,6 +1163,7 @@
         <property name="visible">True</property>
         <property name="can-focus">False</property>
         <property name="no-show-all">True</property>
+        <property name="visible-vertical">False</property>
         <child>
           <object class="GtkMenuButton" id="btn-advanced">
             <property name="visible">True</property>
@@ -1181,6 +1188,7 @@
         <property name="visible">True</property>
         <property name="can-focus">False</property>
         <property name="no-show-all">True</property>
+        <property name="visible-vertical">False</property>
         <child>
           <object class="GtkMenuButton" id="btn-simple">
             <property name="visible">True</property>
@@ -1200,5 +1208,410 @@
         <property name="homogeneous">False</property>
       </packing>
     </child>
+    <child>
+      <object class="GtkSeparatorToolItem">
+        <property name="visible">True</property>
+        <property name="can-focus">False</property>
+        <property name="visible-horizontal">False</property>
+      </object>
+      <packing>
+        <property name="expand">False</property>
+        <property name="homogeneous">True</property>
+      </packing>
+    </child>
+    <child>
+      <object class="GtkToggleToolButton">
+        <property name="visible">True</property>
+        <property name="can-focus">False</property>
+        <property name="no-show-all">True</property>
+        <property name="visible-horizontal">False</property>
+        <property name="action-name">win.snap-bbox</property>
+        <property name="icon-name">snap-bounding-box</property>
+      </object>
+      <packing>
+        <property name="expand">False</property>
+        <property name="homogeneous">False</property>
+      </packing>
+    </child>
+    <child>
+      <object class="GtkToggleToolButton">
+        <property name="visible">True</property>
+        <property name="can-focus">False</property>
+        <property name="no-show-all">True</property>
+        <property name="visible-horizontal">False</property>
+        <property name="action-name">win.snap-bbox-edge</property>
+        <property name="icon-name">snap-bounding-box-edges</property>
+      </object>
+      <packing>
+        <property name="expand">False</property>
+        <property name="homogeneous">False</property>
+      </packing>
+    </child>
+    <child>
+      <object class="GtkToggleToolButton">
+        <property name="visible">True</property>
+        <property name="can-focus">False</property>
+        <property name="no-show-all">True</property>
+        <property name="visible-horizontal">False</property>
+        <property name="action-name">win.snap-bbox-corner</property>
+        <property name="icon-name">snap-bounding-box-corners</property>
+      </object>
+      <packing>
+        <property name="expand">False</property>
+        <property name="homogeneous">False</property>
+      </packing>
+    </child>
+    <child>
+      <object class="GtkToggleToolButton">
+        <property name="visible">True</property>
+        <property name="can-focus">False</property>
+        <property name="no-show-all">True</property>
+        <property name="visible-horizontal">False</property>
+        <property name="action-name">win.snap-bbox-edge-midpoint</property>
+        <property name="icon-name">snap-bounding-box-midpoints</property>
+      </object>
+      <packing>
+        <property name="expand">False</property>
+        <property name="homogeneous">False</property>
+      </packing>
+    </child>
+    <child>
+      <object class="GtkToggleToolButton">
+        <property name="visible">True</property>
+        <property name="can-focus">False</property>
+        <property name="no-show-all">True</property>
+        <property name="visible-horizontal">False</property>
+        <property name="action-name">win.snap-bbox-center</property>
+        <property name="icon-name">snap-bounding-box-center</property>
+      </object>
+      <packing>
+        <property name="expand">False</property>
+        <property name="homogeneous">False</property>
+      </packing>
+    </child>
+    <child>
+      <object class="GtkSeparatorToolItem">
+        <property name="visible">True</property>
+        <property name="can-focus">False</property>
+        <property name="visible-horizontal">False</property>
+      </object>
+      <packing>
+        <property name="expand">False</property>
+        <property name="homogeneous">True</property>
+      </packing>
+    </child>
+    <child>
+      <object class="GtkToggleToolButton">
+        <property name="visible">True</property>
+        <property name="can-focus">False</property>
+        <property name="no-show-all">True</property>
+        <property name="visible-horizontal">False</property>
+        <property name="action-name">win.snap-node-category</property>
+        <property name="icon-name">snap-nodes</property>
+      </object>
+      <packing>
+        <property name="expand">False</property>
+        <property name="homogeneous">False</property>
+      </packing>
+    </child>
+    <child>
+      <object class="GtkToggleToolButton">
+        <property name="visible">True</property>
+        <property name="can-focus">False</property>
+        <property name="no-show-all">True</property>
+        <property name="visible-horizontal">False</property>
+        <property name="action-name">win.snap-path</property>
+        <property name="icon-name">snap-nodes-path</property>
+      </object>
+      <packing>
+        <property name="expand">False</property>
+        <property name="homogeneous">False</property>
+      </packing>
+    </child>
+    <child>
+      <object class="GtkToggleToolButton">
+        <property name="visible">True</property>
+        <property name="can-focus">False</property>
+        <property name="no-show-all">True</property>
+        <property name="visible-horizontal">False</property>
+        <property name="action-name">win.snap-path-intersection</property>
+        <property name="icon-name">snap-nodes-intersection</property>
+      </object>
+      <packing>
+        <property name="expand">False</property>
+        <property name="homogeneous">False</property>
+      </packing>
+    </child>
+    <child>
+      <object class="GtkToggleToolButton">
+        <property name="visible">True</property>
+        <property name="can-focus">False</property>
+        <property name="no-show-all">True</property>
+        <property name="visible-horizontal">False</property>
+        <property name="action-name">win.snap-node-cusp</property>
+        <property name="icon-name">snap-nodes-cusp</property>
+      </object>
+      <packing>
+        <property name="expand">False</property>
+        <property name="homogeneous">False</property>
+      </packing>
+    </child>
+    <child>
+      <object class="GtkToggleToolButton">
+        <property name="visible">True</property>
+        <property name="can-focus">False</property>
+        <property name="no-show-all">True</property>
+        <property name="visible-horizontal">False</property>
+        <property name="action-name">win.snap-node-smooth</property>
+        <property name="icon-name">snap-nodes-smooth</property>
+      </object>
+      <packing>
+        <property name="expand">False</property>
+        <property name="homogeneous">False</property>
+      </packing>
+    </child>
+    <child>
+      <object class="GtkToggleToolButton">
+        <property name="visible">True</property>
+        <property name="can-focus">False</property>
+        <property name="no-show-all">True</property>
+        <property name="visible-horizontal">False</property>
+        <property name="action-name">win.snap-line-midpoint</property>
+        <property name="icon-name">snap-nodes-midpoint</property>
+      </object>
+      <packing>
+        <property name="expand">False</property>
+        <property name="homogeneous">False</property>
+      </packing>
+    </child>
+    <child>
+      <object class="GtkSeparatorToolItem">
+        <property name="visible">True</property>
+        <property name="can-focus">False</property>
+        <property name="visible-horizontal">False</property>
+      </object>
+      <packing>
+        <property name="expand">False</property>
+        <property name="homogeneous">True</property>
+      </packing>
+    </child>
+    <child>
+      <object class="GtkToggleToolButton">
+        <property name="visible">True</property>
+        <property name="can-focus">False</property>
+        <property name="no-show-all">True</property>
+        <property name="visible-horizontal">False</property>
+        <property name="action-name">win.snap-others</property>
+        <property name="icon-name">snap-others</property>
+      </object>
+      <packing>
+        <property name="expand">False</property>
+        <property name="homogeneous">False</property>
+      </packing>
+    </child>
+    <child>
+      <object class="GtkToggleToolButton">
+        <property name="visible">True</property>
+        <property name="can-focus">False</property>
+        <property name="no-show-all">True</property>
+        <property name="visible-horizontal">False</property>
+        <property name="action-name">win.snap-object-midpoint</property>
+        <property name="icon-name">snap-nodes-center</property>
+      </object>
+      <packing>
+        <property name="expand">False</property>
+        <property name="homogeneous">False</property>
+      </packing>
+    </child>
+    <child>
+      <object class="GtkToggleToolButton">
+        <property name="visible">True</property>
+        <property name="can-focus">False</property>
+        <property name="no-show-all">True</property>
+        <property name="visible-horizontal">False</property>
+        <property name="action-name">win.snap-rotation-center</property>
+        <property name="icon-name">snap-nodes-rotation-center</property>
+      </object>
+      <packing>
+        <property name="expand">False</property>
+        <property name="homogeneous">False</property>
+      </packing>
+    </child>
+    <child>
+      <object class="GtkToggleToolButton">
+        <property name="visible">True</property>
+        <property name="can-focus">False</property>
+        <property name="no-show-all">True</property>
+        <property name="visible-horizontal">False</property>
+        <property name="action-name">win.snap-text-baseline</property>
+        <property name="icon-name">snap-text-baseline</property>
+      </object>
+      <packing>
+        <property name="expand">False</property>
+        <property name="homogeneous">False</property>
+      </packing>
+    </child>
+    <child>
+      <object class="GtkToggleToolButton">
+        <property name="visible">True</property>
+        <property name="can-focus">False</property>
+        <property name="no-show-all">True</property>
+        <property name="visible-horizontal">False</property>
+        <property name="action-name">win.snap-grid</property>
+        <property name="icon-name">grid-rectangular</property>
+      </object>
+      <packing>
+        <property name="expand">False</property>
+        <property name="homogeneous">False</property>
+      </packing>
+    </child>
+    <child>
+      <object class="GtkToggleToolButton">
+        <property name="visible">True</property>
+        <property name="can-focus">False</property>
+        <property name="no-show-all">True</property>
+        <property name="visible-horizontal">False</property>
+        <property name="action-name">win.snap-guide</property>
+        <property name="icon-name">guides</property>
+      </object>
+      <packing>
+        <property name="expand">False</property>
+        <property name="homogeneous">False</property>
+      </packing>
+    </child>
+    <child>
+      <object class="GtkToggleToolButton">
+        <property name="visible">True</property>
+        <property name="can-focus">False</property>
+        <property name="no-show-all">True</property>
+        <property name="visible-horizontal">False</property>
+        <property name="action-name">win.snap-page-border</property>
+        <property name="icon-name">snap-page</property>
+      </object>
+      <packing>
+        <property name="expand">False</property>
+        <property name="homogeneous">False</property>
+      </packing>
+    </child>
+    <child>
+      <object class="GtkSeparatorToolItem">
+        <property name="visible">True</property>
+        <property name="can-focus">False</property>
+        <property name="visible-horizontal">False</property>
+      </object>
+      <packing>
+        <property name="expand">False</property>
+        <property name="homogeneous">True</property>
+      </packing>
+    </child>
+    <child>
+      <object class="GtkToggleToolButton">
+        <property name="visible">True</property>
+        <property name="can-focus">False</property>
+        <property name="no-show-all">True</property>
+        <property name="visible-horizontal">False</property>
+        <property name="action-name">win.snap-alignment</property>
+        <property name="icon-name">snap-alignment</property>
+      </object>
+      <packing>
+        <property name="expand">False</property>
+        <property name="homogeneous">False</property>
+      </packing>
+    </child>
+    <child>
+      <object class="GtkToggleToolButton">
+        <property name="visible">True</property>
+        <property name="can-focus">False</property>
+        <property name="no-show-all">True</property>
+        <property name="visible-horizontal">False</property>
+        <property name="action-name">win.snap-alignment-self</property>
+        <property name="icon-name">snap-alignment-self</property>
+      </object>
+      <packing>
+        <property name="expand">False</property>
+        <property name="homogeneous">False</property>
+      </packing>
+    </child>
+    <child>
+      <object class="GtkToggleToolButton">
+        <property name="visible">True</property>
+        <property name="can-focus">False</property>
+        <property name="no-show-all">True</property>
+        <property name="visible-horizontal">False</property>
+        <property name="action-name">win.snap-path-mask</property>
+        <property name="icon-name">path-mask-edit</property>
+      </object>
+      <packing>
+        <property name="expand">False</property>
+        <property name="homogeneous">False</property>
+      </packing>
+    </child>
+    <child>
+      <object class="GtkToggleToolButton">
+        <property name="visible">True</property>
+        <property name="can-focus">False</property>
+        <property name="no-show-all">True</property>
+        <property name="visible-horizontal">False</property>
+        <property name="action-name">win.snap-path-clip</property>
+        <property name="icon-name">path-clip-edit</property>
+      </object>
+      <packing>
+        <property name="expand">False</property>
+        <property name="homogeneous">False</property>
+      </packing>
+    </child>
+    <child>
+      <object class="GtkSeparatorToolItem">
+        <property name="visible">True</property>
+        <property name="can-focus">False</property>
+        <property name="visible-horizontal">False</property>
+      </object>
+      <packing>
+        <property name="expand">False</property>
+        <property name="homogeneous">True</property>
+      </packing>
+    </child>
+    <child>
+      <object class="GtkToggleToolButton">
+        <property name="visible">True</property>
+        <property name="can-focus">False</property>
+        <property name="no-show-all">True</property>
+        <property name="visible-horizontal">False</property>
+        <property name="action-name">win.snap-distribution</property>
+        <property name="icon-name">snap-distribution</property>
+      </object>
+      <packing>
+        <property name="expand">False</property>
+        <property name="homogeneous">False</property>
+      </packing>
+    </child>
+    <child>
+      <object class="GtkToggleToolButton">
+        <property name="visible">True</property>
+        <property name="can-focus">False</property>
+        <property name="no-show-all">True</property>
+        <property name="visible-horizontal">False</property>
+        <property name="action-name">win.snap-line-perpendicular</property>
+        <property name="icon-name">snap-lines-perpendicular</property>
+      </object>
+      <packing>
+        <property name="expand">False</property>
+        <property name="homogeneous">False</property>
+      </packing>
+    </child>
+    <child>
+      <object class="GtkToggleToolButton">
+        <property name="visible">True</property>
+        <property name="can-focus">False</property>
+        <property name="no-show-all">True</property>
+        <property name="visible-horizontal">False</property>
+        <property name="action-name">win.snap-line-tangential</property>
+        <property name="icon-name">snap-lines-tangential</property>
+      </object>
+      <packing>
+        <property name="expand">False</property>
+        <property name="homogeneous">False</property>
+      </packing>
+    </child>
   </object>
 </interface>
diff --git a/src/ui/dialog/inkscape-preferences.cpp b/src/ui/dialog/inkscape-preferences.cpp
index a607b09b3e..4fb2e6f27e 100644
--- a/src/ui/dialog/inkscape-preferences.cpp
+++ b/src/ui/dialog/inkscape-preferences.cpp
@@ -1876,7 +1876,8 @@ void InkscapePreferences::initPageUI()
 
         std::vector<PrefItem> snap = {
             { _("Simple"), 1, _("Present simplified snapping options that manage all advanced settings"), true },
-            { _("Advanced"), 0, _("Expose all snapping options for manual control") }
+            { _("Advanced"), 0, _("Expose all snapping options for manual control") },
+            { _("Permanent"), 2, _("All advanced snap options appear in a permanent bar") },
         };
         _page_toolbars.add_line(false, _("Snap controls bar:"), *Gtk::make_managed<PrefRadioButtons>(snap, "/toolbox/simplesnap"), "", "");
     } catch (const Glib::Error &ex) {
diff --git a/src/widgets/desktop-widget.cpp b/src/widgets/desktop-widget.cpp
index e2bb7c91df..fa1cf37111 100644
--- a/src/widgets/desktop-widget.cpp
+++ b/src/widgets/desktop-widget.cpp
@@ -200,8 +200,6 @@ SPDesktopWidget::SPDesktopWidget(InkscapeWindow* inkscape_window)
     dtw->aux_toolbox = ToolboxFactory::createAuxToolbox();
 
     dtw->snap_toolbox = ToolboxFactory::createSnapToolbox();
-    ToolboxFactory::setOrientation(dtw->snap_toolbox, GTK_ORIENTATION_HORIZONTAL);
-    dtw->_top_toolbars->attach(*Glib::wrap(dtw->snap_toolbox), 1, 0, 1, 2);
 
     dtw->commands_toolbox = ToolboxFactory::createCommandsToolbox();
     auto cmd = Glib::wrap(dtw->commands_toolbox);
@@ -212,6 +210,10 @@ SPDesktopWidget::SPDesktopWidget(InkscapeWindow* inkscape_window)
     dtw->tool_toolbox = ToolboxFactory::createToolToolbox(inkscape_window);
     ToolboxFactory::setOrientation( dtw->tool_toolbox, GTK_ORIENTATION_VERTICAL );
     dtw->_tbbox->pack1(*Glib::wrap(dtw->tool_toolbox), false, true);
+
+    _tb_snap_pos = prefs->createObserver("/toolbox/simplesnap", sigc::mem_fun(this, &SPDesktopWidget::repack_snaptoolbar));
+    repack_snaptoolbar();
+
     auto tbox_width = prefs->getEntry("/toolbox/tools/width");
     if (tbox_width.isValid()) {
         _tbbox->set_position(tbox_width.getIntLimited(32, 8, 500));
@@ -1268,6 +1270,25 @@ SPDesktopWidget::SPDesktopWidget(InkscapeWindow *inkscape_window, SPDocument *do
     dtw->_panels->setDesktop(dtw->desktop);
 }
 
+/**
+ * Choose where to pack the snap toolbar.
+ */
+void SPDesktopWidget::repack_snaptoolbar()
+{
+    Inkscape::Preferences* prefs = Inkscape::Preferences::get();
+    if (auto widget = Glib::wrap(snap_toolbox)) {
+        if (auto parent = widget->get_parent()) {
+            parent->remove(*widget);
+        }
+        if (prefs->getInt("/toolbox/simplesnap", 1) == 2) {
+            ToolboxFactory::setOrientation(snap_toolbox, GTK_ORIENTATION_VERTICAL);
+            _hbox->pack_end(*widget, false, true);
+        } else {
+            ToolboxFactory::setOrientation(snap_toolbox, GTK_ORIENTATION_HORIZONTAL);
+            _top_toolbars->attach(*Glib::wrap(snap_toolbox), 1, 0, 1, 2);
+        }
+    }
+}
 
 void
 SPDesktopWidget::update_rulers()
diff --git a/src/widgets/desktop-widget.h b/src/widgets/desktop-widget.h
index 87ae57476d..63eb0572c6 100644
--- a/src/widgets/desktop-widget.h
+++ b/src/widgets/desktop-widget.h
@@ -199,6 +199,7 @@ public:
     void update_zoom();
     void update_rotation();
     void update_rulers();
+    void repack_snaptoolbar();
 
     void iconify();
     void maximize();
@@ -218,6 +219,7 @@ private:
     GtkWidget *aux_toolbox;
     GtkWidget *commands_toolbox;
     GtkWidget *snap_toolbox;
+    Inkscape::PrefObserver _tb_snap_pos;
     Inkscape::PrefObserver _tb_icon_sizes1;
     Inkscape::PrefObserver _tb_icon_sizes2;
     Inkscape::PrefObserver _tb_visible_buttons;
diff --git a/src/widgets/toolbox.cpp b/src/widgets/toolbox.cpp
index f449e33004..86f983f662 100644
--- a/src/widgets/toolbox.cpp
+++ b/src/widgets/toolbox.cpp
@@ -292,7 +292,6 @@ GtkWidget *ToolboxFactory::createSnapToolbox()
 {
     auto tb = new SnapBar();
     tb->set_name("SnapToolbox");
-    tb->set_orientation(Gtk::ORIENTATION_VERTICAL);
     tb->set_homogeneous(false);
 
     Glib::ustring snap_toolbar_builder_file = get_filename(UIS, "toolbar-snap.ui");
-- 
GitLab


